### PR TITLE
Add an option to distribute reads among all nodes

### PR DESCRIPTION
**Pull request by o1egl**
In my current setup I have 1 master and 1 slave with 90% of read operations and 10% of write operations. This means that the resources of the host are not being used properly. The idea is to have and option to distribute read requests between the master and slave.

### CODE REVIEW

**Comment by linxGnu** [comment]
Hello @o1egl 
Thank you for your consideration. May I ask you for the purpose/use case clearly?

As your PR shows up, additional option targets the case that we could query to one of masters or one of slaves. The balancer only picks one that succeed. Is it your purpose?

**Comment by o1egl** [comment]
@linxGnu I added a description

File: mssqlx.go [hunk 1]
```diff
@@ -1437,12 +1445,32 @@ func (dbs *DBs) BeginTxx(ctx context.Context, opts *sql.TxOptions) (res *sqlx.Tx
...
+	}
+}
+
 // ConnectMasterSlaves to master-slave databases, healthchecks will ensure they are working
 // driverName: mysql, postgres, etc.
 // masterDSNs: data source names of Masters.
 // slaveDSNs: data source names of Slaves.
-// args: args[0] = true to indicates galera/wsrep cluster.
-func ConnectMasterSlaves(driverName string, masterDSNs []string, slaveDSNs []string, args ...interface{}) (*DBs, []error) {
+func ConnectMasterSlaves(driverName string, masterDSNs []string, slaveDSNs []string, options ...Option) (*DBs, []error) {
```
Comment (linxGnu) [inline mssqlx.go:1473]: I love currying options like these. Thank you!

File: mssqlx.go [hunk 2]
```diff
@@ -1437,12 +1445,32 @@ func (dbs *DBs) BeginTxx(ctx context.Context, opts *sql.TxOptions) (res *sqlx.Tx
...
 
+type clusterArgs struct {
+	isWsrep     bool
+	readFromAll bool
+}
+
+type Option func(*clusterArgs)
+
+// IsWsrep indicates galera/wsrep cluster
+func IsWsrep() Option {
```
Comment (linxGnu) [inline mssqlx.go:1456]: How about:
```
// MarkWsrep indicates ....
func MarkWsrep()
```

File: mssqlx.go [hunk 3]
```diff
@@ -1437,12 +1445,32 @@ func (dbs *DBs) BeginTxx(ctx context.Context, opts *sql.TxOptions) (res *sqlx.Tx
...
+	return func(o *clusterArgs) {
+		o.isWsrep = true
+	}
+}
+
+// ReadFromAll sets flag to send read requests to masters and slaves
+func ReadFromAll() Option {
```
Comment (linxGnu) [inline mssqlx.go:1463]: How about:
```
type ReadQuerySource int

const (
   // Slave indicates: read-queries will be distributed only among slaves.
   // This is default value.
   //
   // Note: there is no option for Master. One could use functions like `QueryMaster`, etc
   // to query from masters only.
   Slaves ReadQuerySource = iota
   // FromAll indicates: read-queries will be distributed among both masters and slaves.
   FromAll
)

// SetReadQuerySource set default sources for read-queries.
func SetReadQuerySource(v ReadQuerySource) {
    ....
}
```

File: mssqlx.go [hunk 4]
```diff
@@ -1437,12 +1445,32 @@ func (dbs *DBs) BeginTxx(ctx context.Context, opts *sql.TxOptions) (res *sqlx.Tx
 	}
 }
 
+type clusterArgs struct {
+	isWsrep     bool
+	readFromAll bool
+}
+
+type Option func(*clusterArgs)
```
Comment (linxGnu) [inline mssqlx.go:1453]: How about:
- Moving Options to another go file
- Update README

File: mssqlx.go [hunk 5]
```diff
@@ -65,6 +66,13 @@ func (dbs *DBs) getDBs(s []*wrapper) ([]*sqlx.DB, int) {
 	return r, n
 }
 
+func (dbs *DBs) getReadBalancer() *balancer {
```
Comment (linxGnu) [inline mssqlx.go:69]: How about:
```
func (dbs *DBs) readBalancer() *balancer {
  // decides by read-queries source
  ...
}
```

File: mssqlx_test.go [hunk 6]
```diff
@@ -454,12 +454,12 @@ func TestConnectMasterSlave(t *testing.T) {
 		t.Fatal("Destroy fail")
 	}
 
-	db, _ = ConnectMasterSlaves(driver, nil, slaveDSNs, true)
+	db, _ = ConnectMasterSlaves(driver, nil, slaveDSNs, IsWsrep())
```
Comment (linxGnu) [inline mssqlx_test.go:457]: Please also set new option here for testing

**Comment by o1egl** [comment]
@linxGnu thank you for your review. I addressed your comments.

**Comment by linxGnu** [comment]
@o1egl Thank you very much for your contribution. It's great to see the PR.
:+1:

**Pull request merged**